### PR TITLE
Replace setup.py with pyproject.toml

### LINKED ISSUE #8090: [python] 'setup.py install' is deprecated
## Description

XGBoost's Python package relies on subclassing some of the commands from `setuptools`, with XGBoost-specific logic to build a shared library.

https://github.com/dmlc/xgboost/blob/e28f6f66577eb31451e5d3fbf4be0a21626f7372/python-package/setup.py#L343-L348

This approach, and specifically support for invocations like `python setup.py install`, is considered deprecated in `setuptools` and may be removed in the future.

See, for example, any recent run of the `XGBoost-Python-Wheels` CI job ([example build link](https://github.com/dmlc/xgboost/runs/7374372443?check_suite_focus=true)).

You'll find a warning like the following.

> python3.8/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.

## Specific Request

I'm not really sure. I noticed this in LightGBM (https://github.com/microsoft/LightGBM/issues/5061), and wanted to document it here as well.

I might attempt to resolve this in LightGBM by trying to switch to [`scikit-build`](https://scikit-build.readthedocs.io/en/latest/), but that isn't something we've worked on or even really discussed in detail yet (https://github.com/microsoft/LightGBM/issues/5061#issuecomment-1066276131).

## References

See much more documentation on this in the description and comments at https://github.com/microsoft/LightGBM/issues/5061.

### PR DESCRIPTION
**Pull request by hcho3**
Closes #8090

* Create `pyproject.toml`
* Implement a custom build backend (see below) in `packager` directory. Build logic from `setup.py` has been refactored and migrated into the new backend.
* Tested: `pip wheel .` (build wheel), `python -m build --sdist .` (source distribution)

TODOs
- [x] Update CI to use new build commands.
- [x] Write a doc for building wheels and sdists. Also show how to set CMake args:
```
pip install -v . --config-setting use_system_libxgboost=True
```
- [x] Handle CMake args in `packager/pep517.py`.
- [x] Test editable installation.
- [x] Add typing annotation to `packager`.

Context
* Directly calling `setup.py` for installing the Python package is deprecated. `python setup.py install` will throw a warning:
```
SetuptoolsDeprecationWarning: setup.py install is deprecated.
Use build and pip and other standards-based tools.
```
* The new replacement for `setup.py` is PEP 517 and PEP 621.
  - PEP 621: The package's metadata is now specified in `pyproject.toml`. The TOML file is declarative and cannot contain arbitrary code unlike `setup.py`. 
  - PEP 517: It is now possible to specify a **build backend** in `pyproject.toml`:
```
[build-system]
requires = ["setuptools"]
build-backend = "setuptools.build_meta"
```

* Users now can choose between a variety of build tools: `setuptools`, `poetry`, `hatch` etc. In addition, individual Python packages can specify a custom (bespoke) build backend by specifying `backend-path`:
```
[build-system]
requires = ["packaging"]
backend-path = ["xgboost"]
build-backend = "packager.pep517"
```
A build backend implements two methods:
  - `def build_wheel(wheel_directory, config_settings=None, metadata_directory=None)`
  - `def build_sdist(sdist_directory, config_settings=None)`

### CODE REVIEW

**Comment by hcho3** [comment]
@jameslamb You might be interested in this work.

**Comment by jameslamb** [comment]
Thanks for tagging me! I'm very interested in this and happy to provide a review or answer questions if you want some extra help.

For reference, in LightGBM I'm pursuing using `scikit-build-core` instead of having a custom build backend: https://github.com/microsoft/LightGBM/pull/5759. But only because I don't know much about the custom build backends. I'll read through the approach you took here.

**Review by trivialfis — COMMENTED** [review]
Just out of curiosity, is it possible that we split the XGBoost package into two like what the conda forge package does? One as libxgboost and another one as (py)xgboost.

CMake is now on [pypi](https://pypi.org/project/cmake/), which has two implications. Firstly it's a C++ project with only a CLI interface. The mentioned libxgboost would be similar: a C++ project with only a C interface. Secondly, I am not sure if there is a distinction between build-time and run-time dependency for toml projects. If so then we can have cmake as a build-time dependency for the source distribution.

**Comment by jameslamb** [comment]
> I am not sure if there is a distinction between build-time and run-time dependency for toml projects

There is! This is one of the nice benefits of the new backends... you don't have to have something like `setuptools` or `wheel` as a runtime dependency.

Also on the topic of CMake being available on PyPI, two very important things to consider:

1. anyone running `pip install cmake` on a platform for which there isn't a `CMake` wheel is going end up compiling it from source, which can be really painful and definitely more difficult than installing it from package managers with broader platform support like `brew` on macOS, `apt` on various Linux flavors
2. because of that, `scikit-build-core` has the behavior "if an already-installed `cmake` is found on `PATH`, use that... otherwise, install the one from PyPI"

**Comment by vyasr** [comment]
> Secondly, I am not sure if there is a distinction between build-time and run-time dependency for toml projects. 

It's worth noting that this is in fact one of the motivating reasons for the creation of the modern build infrastructure, namely PEP 517 as previously mentioned as well as [PEP 518](https://peps.python.org/pep-0518/), which defines the `build-system` part of `pyproject.toml`. The canonical issue this was designed to fix is a package that requires numpy and/or Cython to be installed in order to run setup.py in order to provide numpy headers (`np.get_include`) or Cythonize files, respectively. Quoting the PEP:

> But when a project chooses to use setuptools, the use of an executable file like setup.py becomes an issue. You can’t execute a setup.py file without knowing its dependencies, but currently there is no standard way to know what those dependencies are in an automated fashion without executing the setup.py file where that information is stored. It’s a catch-22 of a file not being runnable without knowing its own contents which can’t be known programmatically unless you run the file.

The `build-system` section of `pyproject.toml` allows specifying the requirements for the build. PEP 517 builds occur in a self-contained virtual environment where all build requirements are installed. Run requirements are a separate list that is installed into the user's environment along with the built package after a build is complete (when using `pip install .` to build locally) or a pre-built binary is found (`pip install ${pkg}`). The contained build is analogous to how conda builds work (in venvs instead of conda envs).

**Comment by trivialfis** [comment]
It's not necessary to be useful, just a way for us to split c++ packaging from the Python package.

**Comment by hcho3** [comment]
@trivialfis Let's not split the packaging. It will introduce headaches, since different virtual env managers for Python install native libs in different locations. We don't want to make `libpath.py` even more complicated than it is now.

**Comment by hcho3** [comment]
@vyasr I managed to use `hatchilng` backend (with some customization). As a result, I was able to remove the boilerplate code, such as the metadata generation. Now XGBoost's build backend only contains the necessary customization. Thanks for your advice!

**Comment by hcho3** [comment]
@trivialfis Editable installation is now supported. Run: `pip install -e ."

**Comment by hcho3** [comment]
@jameslamb To give you some context, I chose a happy middle path between implementing a backend from scratch and using an existing backend. I built on top of `hatchling`, which led to modular code with minimal amount of boilerplate.

**Review by trivialfis — COMMENTED** [review]
Thank you for working on the Python build system @hcho3 , I don't have enough expertise to make technical comments. It would be great if you can update the document for the new build system https://github.com/dmlc/xgboost/blob/master/doc/build.rst#building-python-package-from-source .

**Comment by hcho3** [comment]
Good news, I am able to remove `insert_vcomp140.py` and fold it into the new build backend.

**Comment by trivialfis** [comment]
> Good news, I am able to remove insert_vcomp140.py and fold it into the new build backend.

We might want to make it optional. I'm not sure what's the implication of doing that for systems like conda.

**Comment by hcho3** [comment]
@trivialfis Good point, let me add a build option then.

**Comment by hcho3** [comment]
I am removing `vcomp140.dll` from the XGBoost wheel. I just tested the wheel on a minimal installation of Windows, and `import xgboost` fails with

> Error message(s): ["Could not find module 'C:\\Users\\Administrator\\Downloads\\foobar\\lib\\site-packages\\xgboost\\lib\\xgboost.dll' (or one of its dependencies). Try using the full path with constructor syntax."]

I tried a couple things to salvage the error, but none worked.

* Preload dependent DLLs using [the approach taken by scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/66a4d9639e27bc04c21bac4eb259fe28b1dad4af/build_tools/github/vendor.py#L48). The approach only works with compiled extension but not vanilla C lib that's loaded via `ctypes.cdll.LoadLibrary()`.
* Use `delvewheel` to patch the wheel. The import still failed.
* Pass `winmode` arg to `ctypes.CDLL`. No avail.
* I tested both the new wheel and 1.7.5 version. All failed.

Conclusion: Let's not bundle `vcomp140.dll` at all, since it doesn't work. Instead, let's tell Windows users to install Visual C++ Redistributable package. This is what LightGBM does. Alternatively, they can use Conda which handles this automatically.

**Comment by trivialfis** [comment]
note to myself: need to update the release script.

**Comment by hcho3** [comment]
@trivialfis I updated `dev/release-artifacts.py`. Can you take a look?

File: tests/ci_build/lint_python.py
Comment (trivialfis) [inline tests/ci_build/lint_python.py:223]: maybe merge this with `python-package/xgboost` and just call mypy with `python-package`?

**Review by trivialfis — APPROVED** [review]
The change looks good to me. Thank you again for working on this! I'm not entirely sure how all the tools interact with each other. Seems complicated as there's code generation during build/install. Would be great if others can take a look as well.

File: tests/ci_build/lint_python.py
Comment (hcho3) [reply tests/ci_build/lint_python.py:223]: @trivialfis I thought you didn't want to distribute `packager` as part of the Python wheel?

Comment (trivialfis) [reply tests/ci_build/lint_python.py:223]: Apologies for the ambiguity, I meant running the `mypy` on `xgboost/python-package` on the CI, not distributing `packager`  with xgboost.


https://github.com/dmlc/xgboost/blob/191d0aa5cfac87cfc122ebbe8b14da4177309d13/tests/ci_build/lint_python.py#L201

Comment (hcho3) [reply tests/ci_build/lint_python.py:223]: If `packager` is not under directory `xgboost/python-package`, we'd need to specify `packager` as a separate entry. But if we move `packager` to `xgboost/python-package`, then we'd be distributing `packager` with XGBoost.

I prefer the first alternative.

Comment (hcho3) [reply tests/ci_build/lint_python.py:223]: My bad, it's sufficient to call `mypy python-package/`

File: python-package/packager/pep517.py
Comment (vyasr) [inline python-package/packager/pep517.py:76]: You should be able to write this a little more neatly using textwrap.dedent and a multiline string to get the desired formatting
```suggestion
    template = textwrap.dedent("""
        from hatchling.builders.hooks.plugin.interface import BuildHookInterface
        class CustomBuildHook(BuildHookInterface):
            def initialize(self, version, build_data):
                build_data['tag'] = '{tag}'"""
    )
```

File: python-package/packager/pep517.py
Comment (vyasr) [inline python-package/packager/pep517.py:80]: Could skip the format by using an f-string above instead of embedding a format specifier `{tag}` to be replaced later.

File: python-package/packager/pep517.py
Comment (vyasr) [inline python-package/packager/pep517.py:180]: More of a future problem than anything worth dealing with now, but in line with my earlier suggestion to split non-CMake and CMake config options it may eventually be worth splitting the config settings that are for the Python vs C++ build since the former may be supportable for editable builds too.

File: python-package/packager/pep517.py
Comment (vyasr) [inline python-package/packager/pep517.py:161]: IMHO it seems kind of arbitrary to define the copying of the cpp subtree in a separate function in another module when the lines above this are all hardcoding paths to copy anyway. I would personally prefer to see this function inlined, but up to you.

File: python-package/packager/nativelib.py
Comment (vyasr) [inline python-package/packager/nativelib.py:142]:
```suggestion
            raise RuntimeError(
```
or maybe a ValueError?

File: tests/buildkite/build-cpu-arm64.sh
Comment (vyasr) [inline tests/buildkite/build-cpu-arm64.sh:21]: Minor note, since you're using `pip wheel --no-deps` everywhere you may be able to use `python -m build -w` as a drop-in replacement for the most part.

Comment (vyasr) [reply tests/buildkite/build-cpu-arm64.sh:21]: You may also be able to use that in your documentation and CI scripts, but I don't have a strong preference either way there.

File: python-package/packager/pep517.py
Comment (vyasr) [inline python-package/packager/pep517.py:69]: Rather than writing this file as a temporary, can't you just commit the `hatch_build.py` file to the repo with the `get_tag` function in that file to be used by the `initialize` function. Then the file would always be present and it would be copied into the temporary dir in `build_wheel` too, right?

**Review by vyasr — CHANGES_REQUESTED** [review]
This looks good! I'm not too familiar with hatch, so I needed to dig a bit through that backend to review some of the more internal components, but this approach seems pretty reasonable to me.

File: doc/contrib/python_packaging.rst
Comment (hcho3) [reply doc/contrib/python_packaging.rst:12]: @vyasr This document is specifically written for XGBoost developers. So running `pip wheel` is expected to produce a binary wheel.

File: doc/contrib/python_packaging.rst
Comment (hcho3) [reply doc/contrib/python_packaging.rst:1]: This document is not meant for general audience. It's specifically meant for XGBoost developers. Let me revise the title to `How to package XGBoost Python package` to make this clear.

Comment (hcho3) [reply doc/contrib/python_packaging.rst:1]: I'll revise other parts of this document so that it's more geared towards packaging XGBoost specifically.

File: doc/contrib/python_packaging.rst
Comment (hcho3) [reply doc/contrib/python_packaging.rst:15]: I should remove the "binary" bit from this sentence here, since the sentence applies generally to all Python packages.

I'll also add a sentence somewhere saying that, for XGBoost specifically, `pip wheel` produces a binary wheel due to the existence of a compiled binary.

File: python-package/packager/build_config.py
Comment (hcho3) [reply python-package/packager/build_config.py:52]: We don't plan to add more CMake flags (or any flags) to the build configuration. The existing CMake flags were moved over due to legacy uses. XGBoost core supports many more CMake flags, and the best way for a developer to use these flags is to invoke CMake manually inside a `build/` directory.

File: doc/install.rst
Comment (hcho3) [reply doc/install.rst:23]: XGBoost requires an in-tree build, which became the default since Pip 21.3+.

From Pip's changelog:

> 21.3 (2021-10-11)
> Deprecations and Removals
> * In-tree builds are now the default. --use-feature=in-tree-build is now ignored. --use-deprecated=out-of-tree-build may be used temporarily to ease the transition. ([#10495](https://github.com/pypa/pip/issues/10495))

File: python-package/packager/build_config.py
Comment (hcho3) [reply python-package/packager/build_config.py:32]: Good point, we should probably throw an error if an unknown setting is passed in.

File: python-package/packager/build_config.py
Comment (hcho3) [reply python-package/packager/build_config.py:55]: > all the getattr usage in this file is probably unnecessary since the attributes will always exist

No, I'm not checking the existence of each attribute. I am simply testing each attribute's Boolean value for True/False. Let me revise the statement to
```python
cmake_value = "ON" if getattr(self, field_name) == True else "OFF"
```
to make it clear.

File: python-package/packager/nativelib.py
Comment (hcho3) [reply python-package/packager/nativelib.py:103]: Some systems (like MacOS) don't have OpenMP installed by default. So in this case, installing XGBoost from the source should fall back to building with OpenMP turned off.

File: python-package/packager/pep517.py
Comment (hcho3) [reply python-package/packager/pep517.py:161]: I put it as a separate module, so that we can easily add more subdirectories for C++ code in the future. If we put the copying logic in `pep517.py`, it would give us more cognitive overhead when we add more C++ files in the future.

File: python-package/packager/nativelib.py
Comment (hcho3) [reply python-package/packager/nativelib.py:65]: The `build_tool` parameter is actually not needed, since the logic for choosing the build tool can be brought into `_build()` function. Let me update the function.

Comment (hcho3) [reply python-package/packager/nativelib.py:65]: As for Windows, a typical setup doesn't have Ninja or Make available as a simple system command. (I think Visual Studio has Ninja and Make somewhere in its installation, but you can't access them easily.)

File: python-package/packager/nativelib.py
Comment (hcho3) [reply python-package/packager/nativelib.py:76]: No, we want to hardcore the list because we have not tested other generators for the Windows platform.

File: python-package/packager/pep517.py
Comment (hcho3) [reply python-package/packager/pep517.py:69]: Good point, let me do that.

File: python-package/packager/pep517.py
Comment (hcho3) [reply python-package/packager/pep517.py:180]: I don't think we want to add more options any time soon.

File: CMakeLists.txt
Comment (hcho3) [reply CMakeLists.txt:277]: `set_output_directory()` is a custom CMake function, which does change the output direcotyr of the final library:
https://github.com/dmlc/xgboost/blob/ef13dd31b1dc073f4732544739db5a252ae5b6e5/cmake/Utils.cmake#L56-L73

As for the reason for putting `libxgboost.so` under the source directory, this is due to legacy reasons. Other language bindings expect `libxgboost.so` to be under `lib/` subdirectory of the source directory.

File: python-package/packager/nativelib.py
Comment (hcho3) [reply python-package/packager/nativelib.py:33]: I personally like using keyword-only arguments, because I am being explicitly about which arguments correspond to which argument. Do you feel that this adds complexity?

File: python-package/packager/nativelib.py
Comment (hcho3) [reply python-package/packager/nativelib.py:93]: I will add a terminating `else` clause.

File: python-package/packager/pep517.py
Comment (hcho3) [reply python-package/packager/pep517.py:25]: Nice, let me add a note about Python 3.11.

File: doc/install.rst
Comment (vyasr) [reply doc/install.rst:23]: Got it. Looks like it can also be done optionally as of 21.1, but no need to change the requirement for that reason.

File: python-package/packager/build_config.py
Comment (vyasr) [reply python-package/packager/build_config.py:51]: Sounds good.

File: python-package/packager/build_config.py
Comment (vyasr) [reply python-package/packager/build_config.py]: Ah you are right, I don't know why I was thinking this had anything to do with attributes being none rather than simply accessing them by name.

File: python-package/packager/nativelib.py
Comment (vyasr) [reply python-package/packager/nativelib.py:33]: I don't mind keyword-only arguments, but I find them odd when they come without defaults. More importantly, in this case I'm not sure why you would have one positional argument and then have two keyword-only arguments with no defaults. That's the main thing that makes for a confusing API IMO. I would make them all keyword-only if that's your preference.

File: CMakeLists.txt
Comment (vyasr) [reply CMakeLists.txt:277]: I see, OK if that is needed for legacy use cases that is fine.

File: python-package/packager/nativelib.py
Comment (vyasr) [reply python-package/packager/nativelib.py:65]: Yep I realized that it was being captured in the closure of the nested function but I was confused about why the build tool could be specified on Linux but not Windows. I know little about Windows though so that seems like a good reason to me.

File: python-package/packager/nativelib.py
Comment (vyasr) [reply python-package/packager/nativelib.py:76]: Got it.

File: python-package/packager/nativelib.py
Comment (vyasr) [reply python-package/packager/nativelib.py:105]: Got it.

File: python-package/packager/pep517.py
Comment (vyasr) [reply python-package/packager/pep517.py:147]: OK that's fine then.

File: python-package/packager/pep517.py
Comment (vyasr) [reply python-package/packager/pep517.py:128]: OK.

File: doc/contrib/python_packaging.rst
Comment (vyasr) [inline doc/contrib/python_packaging.rst:80]:
```suggestion
.. rubric:: Footnotes
```

File: python-package/packager/pep517.py
Comment (vyasr) [inline python-package/packager/pep517.py:153]: Out of curiosity, what happens if you remove this? Is it possible to use an editable installation that also builds libxgboost? In that situation running `pip install -e` would need to rebuild the library as well.

Not saying this feature should be implemented, just want to understand the current setup of the backend to figure out how best to work with this in the future for other RAPIDS libraries too.

File: doc/contrib/python_packaging.rst
Comment (vyasr) [reply doc/contrib/python_packaging.rst:12]: Understood.

**Review by vyasr — APPROVED** [review]
This is looking good! There are a couple of outstanding comments, but almost everything has been addressed and nothing is left that I would block merging on. Great to see this work here!

**Comment by vyasr** [comment]
Note that I don't have permissions to resolve threads on this PR so I tried to leave comments or clearly indicate where I thought things were resolved.

File: python-package/packager/pep517.py
Comment (hcho3) [reply python-package/packager/pep517.py:153]: > what happens if you remove this?

Editable installation won't build `libxgboost.so`, so `import xgboost` will error out.

> Is it possible to use an editable installation that also builds libxgboost?

I decided against it, as there is no reliable way for `packager` module to detect changes in the C++ code. If we want to re-build `libxgboost.so` upon changes made to C++, `scikit-build-core` will probably be a better choice.

I ultimately decided against `scikit-build-core` because:
* It doesn't yet support editable installation.
* I couldn't customize it enough to fit the current use cases of XGBoost.
* Much of what `scikit-learn-core` offers is not useful for XGBoost, since XGBoost does not use compiled extensions. XGBoost uses `ctypes` exclusively to access C API functions from `libxgboost.so`.

The calculus will be different for RAPIDS libraries which uses compiled extensions extensively. IMHO, if you use compiled extensions a lot, it's better to use `scikit-build-core` and give up on editable installation.

File: doc/contrib/python_packaging.rst
Comment (hcho3) [reply doc/contrib/python_packaging.rst:80]: Good catch!

File: python-package/packager/pep517.py
Comment (jameslamb) [reply python-package/packager/pep517.py]: For what it's worth, `scikit-build-core` just added experimental support for editable installs: https://github.com/scikit-build/scikit-build-core/releases/tag/v0.3.0

Not sure exactly what "experimental" means, to be fair, and to what extent that's acceptable for a project as widely-used as XGBoost. But just wanted to note it.

I am planning to pursue `scikit-build-core` for LightGBM (https://github.com/microsoft/LightGBM/pull/5759), to not have to have any lingering Python glue score that's exposed to the changes in Python packaging. That is still very much a work in progress though, so I'm definitely not recommending that you switch to that over `hatchling`... just noting the link as something you may want to watch as we all figure out the `setup.py`-free world together 😊

File: python-package/packager/nativelib.py
Comment (hcho3) [reply python-package/packager/nativelib.py:32]: I removed all `*` from function signatures.

File: python-package/packager/pep517.py
Comment (hcho3) [reply python-package/packager/pep517.py]: @jameslamb Guess my knowledge was outdated :) In the future, I'll definitely consider `scikit-learn-core` as a potential alternative.

Comment (vyasr) [reply python-package/packager/pep517.py]: I will be migrating RAPIDS to use `scikit-build-core` at some point in the not-too-distant future. We're already using `scikit-build`, so the transition is the most natural way to get us onto a PEP 517 build. 

Support for editable installs is a must for RAPIDS, and is one of the reasons I haven't attempted the migration yet. [That feature request is something they're aware of](https://github.com/scikit-build/scikit-build-core/issues/114) and are working on. [My understanding is that the current experimental support](https://github.com/scikit-build/scikit-build-core/pull/212) only covers some possible cases around library changes but not all. I will definitely be keeping a close eye on this. I expect full support for editable installs is a precondition of a 1.0 release, and the project has been moving at a reasonable pace so it shouldn't be too long before that exists.

That said, at this stage `scikit-build-core` is definitely less mature than hatchling and I can't fault your choice since you seem to have been able to get everything working fairly painlessly :slightly_smiling_face:

Comment (jameslamb) [reply python-package/packager/pep517.py]: @vyasr @hcho3 @trivialfis FYI I have a PR ready over in LightGBM for using `scikit-build-core`: https://github.com/microsoft/LightGBM/pull/5759. If you have time and interested, I'd appreciate any comments you have on it.

We don't care about editable installs for `lightgbm`, and I'm willing to go contribute to `scikit-build-core` (both code and discussions) as we discover issues in it, to help that tool mature.

Comment (vyasr) [reply python-package/packager/pep517.py]: Thanks for the ping! I've started testing the migration for a couple RAPIDS libraries as well, see https://github.com/rapidsai/rmm/pull/1287 and https://github.com/rapidsai/cudf/pull/13531. Editable installs are already in significantly better shape than they were a couple of months ago FWIW.

Comment (jameslamb) [reply python-package/packager/pep517.py]: Nice! I learned a lot about `scikit-build-core` in this process, `@` me any time on those RAPIDS PRs if you want an opinion or another set of eyes.

**Pull request merged**
